Compile error when using multiple nested generic struct & generic method

<!-- Please make sure to run `v up` before reporting any issues as it may have already been fixed.
     It's also advisable to update all relevant modules using `v outdated` and `v install` -->

<!-- You can use `v doctor` to fill up the next fields -->
**V version:** V 0.3.0 1c6366e
**OS:** macos, macOS, 12.5, 21G72

<!-- Please include the standalone minimal reproducible code.
If not, it may be closed by the administrators. -->
**What did you do?**
```v
struct Wrapper<S> {
mut:
	calc Calc<S>
}
fn (mut w Wrapper<S>) next<T>(input T) f64 {
	$if S is TypeA || S is TypeB || S is TypeC {
		$if T is f64 {
			return w.calc.next(input)
		} $else $if T is SecretCase {
			return w.calc.next(input)
		} $else {
			panic('"$T.name" is not supported')
			return -1
		}
		return -1
	} $else {
		panic('"$S.name" is not supported')
		return -1
	}
}
struct Calc<S> {
mut:
	typ S
}

struct TypeA {}
struct TypeB {}
struct TypeC {}
interface SecretCase {
	secret() f64
}
interface SpecialCase {
	SecretCase
}
struct SecretSkill {}
fn (s &SecretSkill) secret() f64 {
	return 101.0
}
fn (mut c Calc<S>) next<T>(input T) f64 {
	$if S is TypeA || S is TypeB || S is TypeC {
		$if T is f64 {
			return c.typ.next(input)
		} $else $if T is int {
			return c.typ.next(input)
		} $else $if T is SpecialCase {
			return c.typ.next(input)
		} $else {
			panic('Unsupported type $T.name')
		}
	} $else {
		panic('Unsupported type $S.name')
	}
}
fn (mut t TypeA) next<T>(input T) f64 {
	return 10
}
fn (mut t TypeB) next<T>(input T) f64 {
	return 11
}
fn (mut t TypeC) next<T>(input T) f64 {
	return 12
}
fn new<S>() Wrapper<S> {
	$if S is TypeA {
		return Wrapper<TypeA> {calc: Calc<TypeA> { typ: TypeA {} }}
	} $else $if S is TypeB {
		return Wrapper<TypeB> {calc: Calc<TypeB> { typ: TypeB {} }}
	} $else $if S is TypeC {
		return Wrapper<TypeC> {calc: Calc<TypeC> { typ: TypeC {} }}
	} $else {
		panic('unknown type $S.name')
	}
}

fn main() {
	{
		mut c := Calc<TypeA> {
			typ: TypeA {}
		}
		assert c.next(100) == 10.0
	}
	{
		mut c := Calc<TypeB> {
			typ: TypeB {}
		}
		assert c.next(100) == 11.0
	}
	{
		mut c := new<TypeA>()
		assert c.next(100) == 10.0
	}
	{
		mut c := new<TypeB>()
		assert c.next(11.1) == 11.0
	}
	{
		mut c := new<TypeC>()
		assert c.next(12.1) == 12.0
	}
	println('OK!!')
}
```

**What did you expect to see?**
OK
 
**What did you see instead?**
```
==================
   ^~~~~~~~~~~~~~~
/tmp/v_502/comptime_generic_multiple_case.14247616884717959960.tmp.c:12356:14: error: implicit declaration of function 'main__Calc_T_main__TypeC_next_T_main__TypeB_f64' is invalid in C99 [-Werror,-Wimplicit-function-declaration]
                        f64 _t3 = main__Calc_T_main__TypeC_next_T_main__TypeB_f64(&w->calc, input);
                                  ^
/tmp/v_502/comptime_generic_multiple_case.14247616884717959960.tmp.c:12356:14: note: did you mean 'main__Calc_T_main__TypeB_next_T_main__TypeB_f64'?
/tmp/v_502/comptime_generic_multiple_case.14247616884717959960.tmp.c:1932:21: note: 'main__Calc_T_main__TypeB_next_T_main__TypeB_f64' declared here
VV_LOCAL_SYMBOL f64 main__Calc_T_main__TypeB_next_T_main__TypeB_f64(main__Calc_T_main__TypeB* c, f64 input);
                    ^
1 warning and 1 error generated.
...
==================
(Use `v -cg` to print the entire error message)

builder error: 
==================
C error. This should never happen.
```

** NOTE **
There is no build error if change declaration of `new` to the following.
```v
fn new<S>() Calc<S> {
	$if S is TypeA {
		return Calc<TypeA> { typ: TypeA {} }
	} $else $if S is TypeB {
		return Calc<TypeB> { typ: TypeB {} }
	} $else $if S is TypeC {
		return Calc<TypeC> { typ: TypeC {} }
	} $else {
		panic('unknown type $S.name')
	}
}
```

--- a/vlib/v/checker/check_types.v
+++ b/vlib/v/checker/check_types.v
@@ -621,26 +621,39 @@ pub fn (mut c Checker) infer_fn_generic_types(func ast.Fn, mut node ast.CallExpr
 		mut typ := ast.void_type
 		for i, param in func.params {
 			// resolve generic struct receiver
-			if node.is_method && param.typ.has_flag(.generic) {
-				sym := c.table.final_sym(node.receiver_type)
-				match sym.info {
-					ast.Struct, ast.Interface, ast.SumType {
-						if !isnil(c.table.cur_fn) && c.table.cur_fn.generic_names.len > 0 { // in generic fn
-							if gt_name in c.table.cur_fn.generic_names
-								&& c.table.cur_fn.generic_names.len == c.table.cur_concrete_types.len {
-								idx := c.table.cur_fn.generic_names.index(gt_name)
-								typ = c.table.cur_concrete_types[idx]
-							}
-						} else { // in non-generic fn
+			if node.is_method && i == 0 && param.typ.has_flag(.generic) {
+				sym := c.table.final_sym(node.left_type)
+				if node.left_type.has_flag(.generic) {
+					match sym.info {
+						ast.Struct, ast.Interface, ast.SumType {
 							receiver_generic_names := sym.info.generic_types.map(c.table.sym(it).name)
-							if gt_name in receiver_generic_names
-								&& sym.info.generic_types.len == sym.info.concrete_types.len {
+							if gt_name in receiver_generic_names {
 								idx := receiver_generic_names.index(gt_name)
-								typ = sym.info.concrete_types[idx]
+								typ = sym.info.generic_types[idx]
 							}
 						}
+						else {}
 					}
-					else {}
+				} else {
+					match sym.info {
+						ast.Struct, ast.Interface, ast.SumType {
+							if !isnil(c.table.cur_fn) && c.table.cur_fn.generic_names.len > 0 { // in generic fn
+								if gt_name in c.table.cur_fn.generic_names
+									&& c.table.cur_fn.generic_names.len == c.table.cur_concrete_types.len {
+									idx := c.table.cur_fn.generic_names.index(gt_name)
+									typ = c.table.cur_concrete_types[idx]
+								}
+							} else { // in non-generic fn
+								receiver_generic_names := sym.info.generic_types.map(c.table.sym(it).name)
+								if gt_name in receiver_generic_names
+									&& sym.info.generic_types.len == sym.info.concrete_types.len {
+									idx := receiver_generic_names.index(gt_name)
+									typ = sym.info.concrete_types[idx]
+								}
+							}
+						}
+						else {}
+					}
 				}
 			}
 			arg_i := if i != 0 && node.is_method { i - 1 } else { i }

--- a/vlib/v/tests/generics_with_multi_nested_generic_method_call_test.v
+++ b/vlib/v/tests/generics_with_multi_nested_generic_method_call_test.v
@@ -1,0 +1,68 @@
+struct Wrapper<S> {
+mut:
+	calc Calc<S>
+}
+
+fn (mut w Wrapper<S>) next<T>(input T) f64 {
+	$if S is TypeA || S is TypeB {
+		$if T is f64 {
+			return w.calc.next(input)
+		} $else {
+			panic('"$T.name" is not supported')
+			return -1
+		}
+		return -1
+	} $else {
+		panic('"$S.name" is not supported')
+		return -1
+	}
+}
+
+struct Calc<S> {
+mut:
+	typ S
+}
+
+struct TypeA {}
+
+struct TypeB {}
+
+fn (mut c Calc<S>) next<T>(input T) f64 {
+	$if S is TypeA || S is TypeB {
+		$if T is f64 {
+			return c.typ.next(input)
+		} $else {
+			panic('Unsupported type $T.name')
+		}
+	} $else {
+		panic('Unsupported type $S.name')
+	}
+}
+
+fn (mut t TypeA) next<T>(input T) f64 {
+	return 10
+}
+
+fn (mut t TypeB) next<T>(input T) f64 {
+	return 11
+}
+
+fn test_generics_with_multi_nested_generic_method_call() {
+	{
+		mut c := Wrapper<TypeA>{
+			calc: Calc<TypeA>{
+				typ: TypeA{}
+			}
+		}
+		assert c.next(100.0) == 10.0
+	}
+	{
+		mut c := Wrapper<TypeB>{
+			calc: Calc<TypeB>{
+				typ: TypeB{}
+			}
+		}
+		assert c.next(100.0) == 11.0
+	}
+	println('OK!!')
+}

